packages/reactivity/src/computed.ts

```diff
@@ -20,6 +20,50 @@ export interface WritableComputedOptions<T> {
   set: ComputedSetter<T>
 }
 
+class _ComputedRef<T> {
+  private _value!: T
+  private _dirty = true
+
+  public readonly effect: ReactiveEffect<T>;
+
+  public [ReactiveFlags.IS_READONLY]: boolean
+
+  constructor(
+    getter: ComputedGetter<T>,
+    private readonly _setter: ComputedSetter<T>,
+    isReadonly: boolean
+  ) {
+    this.effect = effect(getter, {
+      lazy: true,
+      scheduler: () => {
+        if (!this._dirty) {
+          this._dirty = true
+          trigger(toRaw(this), TriggerOpTypes.SET, 'value')
+        }
+      }
+    })
+
+    this[ReactiveFlags.IS_READONLY] = isReadonly
+  }
+
+  get value() {
+    if (this._dirty) {
+      this._value = this.effect()
+      this._dirty = false
+    }
+    track(toRaw(this), TrackOpTypes.GET, 'value')
+    return this._value
+  }
+
+  set value(newValue: T) {
+    this._setter(newValue)
+  }
+
+  get __v_isRef() {
```

Comment:
Would it be better to make this a plain property (which would make `isRef` checks faster)?

I get that the getter also ensures it's not writable but I don't think anyone's actually gonna do that.

Reply:
afaik plain property can be readonly too

Reply:
It would require explicit `Object.defineProperty` call though.

Reply:
Would be worth benchmarking the difference. 
As `isRef` is surely not gonna be a monomorphic call site, it's most likely gonna be slow in both cases (I'd bet on the field being slightly faster, prob. not significantly). 
If the browser somehow manages to create a path optimized for `Ref` instances, the getter is likely to be faster for the same reason as `value` is. Being on the prototype with trivial implementation, it might even compile to a constant in this case.

Making it a field is eating up the memory benefit, though.

Thinking this further: would we be ok to expose the `Ref` class as public implementation?
If we do, we could completely drop the `isRef` API in exchange of `instanceof Ref`, which is more likely to be optimized call sites (which are likely to be monomorphic, unlike `isRef` that is called from everywhere). That requires some benchmarking obviously.

Furthermore, we could also drop the `customRef` API, in exchange of `class extends Ref`. The resulting custom refs would be simpler to write and more performant. (This change requires putting a `trigger` and `track` on the `Ref` prototype, which is totally doable).

Reply:
Hm indeed, already changed this to a field but in terms of memory, we may need to revert to a getter.

Reply:
Interesting. Maybe we can bench for memory between a field vs getter?

Exposing `Ref` class directly is a major API change which we are not likely going to consider at this point, and it brings along all the class-related assumptions (extensions, `instanceof` checks) that we do not want to commit to. And I doubt the performance gain would be significant to warrant it (it might be visible in ref-only benchmarks but unlikely to matter in actual applications)

Reply:
Here we go: https://gist.github.com/RobbinBaauw/dff784f226950e888a64c39febd927d0.
For me it was about a 4MB difference.

A larger difference arrives in the performance of `isRef` though: https://jsbench.me/cike1jpqnw. I get a 20% difference in favor of the field. Of course this will fluctuate based on what you pass, but I pass a combination of refs and non-refs so I think it is somewhat representative.